balancer-js/src/modules/pools/apr/apr.ts

```diff
@@ -0,0 +1,332 @@
+              subPool,
+              this.tokenPrices,
+              this.tokenMeta,
+              this.pools,
+              this.liquidityGauges,
+              this.feeDistributor,
+              this.feeCollector,
+              this.tokenYields
+            ).apr()
+          ).min;
+        }
+      }
+
+      return apr;
+    });
+
+    // Get token weights normalised by usd price
+    const weights = bptFreeTokens.map(async (token): Promise<number> => {
+      if (token.weight) {
+        return parseFloat(token.weight);
+      } else {
+        const tokenPrice =
+          token.price?.usd || (await this.tokenPrices.find(token.address))?.usd;
+        if (!tokenPrice) {
+          throw `No price for ${token.address}`;
+        }
+        // using floats assuming frontend purposes with low precision needs
+        const tokenValue = parseFloat(token.balance) * parseFloat(tokenPrice);
+        return tokenValue / parseFloat(totalLiquidity);
+      }
+    });
+
+    // Normalise tokenAPRs according to weights
+    const weightedAprs = await Promise.all(
+      aprs.map(async (apr, idx) => {
+        const [a, w] = await Promise.all([apr, weights[idx]]);
+        return Math.round(a * w);
+      })
+    );
+
+    // sum them up to get pool APRs
+    const apr = weightedAprs.reduce((sum, apr) => sum + apr, 0);
+
+    return apr;
+  }
+
+  /**
+   * Calculates staking rewards based on veBal gauges deployed with Curve Finance contracts.
+   * https://curve.readthedocs.io/dao-gauges.html
+   *
+   * Terminology:
+   *  - LP token of a gauge is a BPT of a pool
+   *  - Depositing into a gauge is called staking on the frontend
+   *  - gauge totalSupply - BPT tokens deposited to a gauge
+   *  - gauge workingSupply - effective BPT tokens participating in reward distribution. sum of 40% deposit + 60% boost from individual user's veBal
+   *  - gauge relative weight - weight of this gauge in bal inflation distribution [0..1] scaled to 1e18
+   *
+   * APR sources:
+   *  - gauge BAL emissions = min: 40% of totalSupply, max: 40% of totalSupply + 60% of totalSupply * gauge LPs voting power
+   *    https://github.com/balancer-labs/balancer-v2-monorepo/blob/master/pkg/liquidity-mining/contracts/gauges/ethereum/LiquidityGaugeV5.vy#L338
+   *  - gauge reward tokens: Admin or designated depositor has an option to deposit additional reward with a weekly accruing cadence.
+   *    https://github.com/balancer-labs/balancer-v2-monorepo/blob/master/pkg/liquidity-mining/contracts/gauges/ethereum/LiquidityGaugeV5.vy#L641
+   *    rate: amount of token per second
+   *
+   * @returns APR [bsp] from protocol rewards.
+   */
+  async stakingApr(boost = 1): Promise<number> {
+    // Data resolving
+    const gauge = await this.liquidityGauges.findBy('poolId', this.pool.id);
+    if (!gauge) {
+      return 0;
+    }
+
+    const balPriceUsd = parseFloat(
+      (
+        await this.tokenPrices.find(
+          '0xba100000625a3754423978a60c9317c58a424e3d'
```

Comment:
This should be a constant somewhere else.

Reply:
it could be. the thing is that it's relevant for mainnet only BAL, because it's related to protocol BAL printer. and even in a testnet case, the price feed would need to be stubbed. I'll just leave it hardcoded.